### PR TITLE
Comment feature

### PR DESCRIPTION
**Pull request by Affaq-Ahmed**
#Description
The Comment and Like Feature have been added to the project. 

##Details:
/comment for the comments routes.
/post/like &
/post/unlike for the like and unlike routes

### CODE REVIEW

File: src/lib/comment.ts [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+		}
+		const reply = new Comment({
+			content,
+			createdBy: user._id,
+			parentCommentId: commentId,
+			postId: comment.postId,
+		});
+		const createdReply = await reply.save();
+
+		await comment.save();
```
Comment (tintash-share) [inline src/lib/comment.ts]: what if this comment failed to save? what is the status of reply than

File: src/lib/comment.ts [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+	res: Response
+) => {
+	const { commentId } = req.params;
+	try {
+		const replies = await Comment.find()
+			.where('parentCommentId')
+			.equals(commentId)
+			.where('deleted')
+			.equals(false)
+			.populate('createdBy', 'name email _id');
```
Comment (tintash-share) [inline src/lib/comment.ts]: try and write JSON queries, because that will help more moving forward

File: src/lib/post.ts [hunk 3]
```diff
@@ -264,4 +262,180 @@ const feed = async (req: IGetUserAuthRequest, res: Response) => {
...
+ * @desc Like a post
+ * @route PUT /post/like/:id
+ * @access Authenticated
+ * @Param postId
+ * @access Moderator can't like post
+ * @access User can't like post if he has no friends
+ */
+const like = async (req: IGetUserAuthRequest, res: Response) => {
+	const { postId } = req.params;
+	const { user } = req;
```
Comment (tintash-share) [inline src/lib/post.ts]: why can't you add these two lines in try catch?

File: src/lib/post.ts [hunk 4]
```diff
@@ -264,4 +262,180 @@ const feed = async (req: IGetUserAuthRequest, res: Response) => {
...
+	try {
+		const post = await Post.findById(postId);
+		if (!post) {
+			logger.info('Post not found');
+			return res.status(400).json({
+				message: 'Post not found',
+			});
+		}
+		if (post.likes?.includes(user._id)) {
```
Comment (tintash-share) [inline src/lib/post.ts]: can we add this in else statement following if above this

File: src/lib/post.ts [hunk 5]
```diff
@@ -264,4 +262,180 @@ const feed = async (req: IGetUserAuthRequest, res: Response) => {
...
+			logger.info('Post already liked');
+			return res.status(400).json({
+				message: 'Post already liked',
+			});
+		}
+		post.likes?.push(user._id);
+		const savedPost = await post.save();
```
Comment (tintash-share) [inline src/lib/post.ts:311]: what's the point saving post if likes never added?

File: src/lib/post.ts [hunk 6]
```diff
@@ -264,4 +262,180 @@ const feed = async (req: IGetUserAuthRequest, res: Response) => {
...
+				message: 'Post not found',
+			});
+		}
+		if (!post.likes?.includes(user._id)) {
+			logger.info('Post not liked');
+			return res.status(400).json({
+				message: 'Post not liked',
+			});
+		}
+		post.likes = post.likes?.filter((like) => like !== user._id);
```
Comment (tintash-share) [inline src/lib/post.ts:352]: what will be assigned to post.likes if post.likes? doesn't exist

File: src/lib/post.ts [hunk 7]
```diff
@@ -264,4 +262,180 @@ const feed = async (req: IGetUserAuthRequest, res: Response) => {
...
+	const { user } = req;
+	try {
+		const post = await Post.findById(id);
+		if (!post) {
+			logger.info('Post not found');
+			return res.status(400).json({
+				message: 'Post not found',
+			});
+		}
+		if (post.deleted) {
```
Comment (tintash-share) [inline src/lib/post.ts:390]: why you are avoiding series of conditions.. like if else if else

File: src/lib/post.ts [hunk 8]
```diff
@@ -264,4 +262,180 @@ const feed = async (req: IGetUserAuthRequest, res: Response) => {
...
+			logger.info('Unauthorized');
+			return res.status(401).json({
+				message: 'Unauthorized',
+			});
+		}
+		const comments = await Comment.find({ postId: id })
+			.where('replyTo')
+			.equals(null);
+
+		// const commentsWithReplies = await Promise.all(
```
Comment (tintash-share) [inline src/lib/post.ts]: remove commented code if not needed

File: src/lib/post.ts [hunk 9]
```diff
@@ -264,4 +262,180 @@ const feed = async (req: IGetUserAuthRequest, res: Response) => {
...
+		// 		const replies = await Comment.find({ parentCommentId: comment._id });
+		// 		return {
+		// 			...comment.toObject(),
+		// 			replies,
+		// 		};
+		// 	})
+		// );
+
+		//loop through comments and add replies to each comment and add replies to each reply and so on
+		const commentsWithReplies = await Promise.all(
```
Comment (tintash-share) [inline src/lib/post.ts]: 1. Move this part to a separate function so it could be reused
2. Why Promise.all is needed?

**Review by tintash-share — COMMENTED** [review]
Please give fixes where needed against comments and resubmit

File: src/lib/post.ts (hunk 5 above)
Comment (Affaq-Ahmed) [reply src/lib/post.ts:311]: I have added the like on line 290. or is it different from what you are saying?

File: src/lib/post.ts (hunk 4 above)
Comment (Affaq-Ahmed) [reply src/lib/post.ts]: yes, we can. but there would be no difference as the response is already being sent in the if block.

File: src/lib/post.ts (hunk 3 above)
Comment (Affaq-Ahmed) [reply src/lib/post.ts]: I had read somewhere or seen in some tutorial that only the DB queries were to be written in the try-catch block. Kindly correct me if I am wrong.

**Review by Affaq-Ahmed — COMMENTED** [review]
I have resubmitted the assignment.

File: src/controllers/comment.ts [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+	commentReplies,
+	createReply,
+} from '../lib/comment';
+
+const createComment = async (req: Request, res: Response) => {
+	const response = await create(req, res);
+};
+
+const getPostComments = async (req: Request, res: Response) => {
+	const response = await postComments(req, res);
```
Comment (abianoor456) [inline src/controllers/comment.ts:15]: Please don't pass req and res to service layer

File: src/lib/post.ts [hunk 11]
```diff
@@ -25,7 +26,7 @@ const getAll = async (req: IGetUserAuthRequest, res: Response) => {
 			posts = await Post.find({})
 				.skip(((page as number) - 1) * (limit as number))
 				.limit(limit as number)
-				.populate('createdBy', '-password -tokens')
+				.populate('createdBy', 'name email')
 				.sort({ createdAt: -1 })
 				.where('deleted')
 				.equals(false);
```
Comment (abianoor456) [inline src/lib/post.ts]: please add filters as json object in find() params

**Review by abianoor456 — CHANGES_REQUESTED** [review]

File: src/controllers/comment.ts (hunk 10 above)
Comment (Affaq-Ahmed) [reply src/controllers/comment.ts:15]: Done

File: src/lib/post.ts (hunk 11 above)
Comment (Affaq-Ahmed) [reply src/lib/post.ts]: Done

**Review by toseefrehan92 — COMMENTED** [review]
LGTM

only thing missing is inline comments, please add to make code self explanatory

File: src/controllers/auth.ts [hunk 12]
```diff
@@ -2,17 +2,33 @@ import { Request, Response } from 'express';
...
+		role = req.body.userRole;
+	} else {
+		role = 'USER';
+	}
+	const response = await userSignup(name, email, password, role);
+
+	return res.status(response.status).json({
+		message: response.message,
+		status: response.status,
+	});
```
Comment (mali-tintash) [inline src/controllers/auth.ts:17]: what happens if response results in an exception?

File: src/controllers/post.ts [hunk 13]
```diff
@@ -7,34 +7,100 @@ import {
...
+	);
+	return res.status(response.status).json(response);
 };
 
-const getFeed = async (req: Request, res: Response) => {
-	const response = await feed(req, res);
+const getFeed = async (req: IGetUserAuthRequest, res: Response) => {
+	const { page = 1, limit = 10, desc = true } = req.query;
+	const { user } = req;
+	if (!user.paid) {
```
Comment (mali-tintash) [inline src/controllers/post.ts:71]: this can read better if we go like 
if (user.paid) {

}
else
{

}

File: src/lib/auth.ts [hunk 14]
```diff
@@ -13,35 +13,38 @@ import { IGetUserAuthRequest } from '../types/Request';
  * @access Public
  * @example http://localhost:3000/users/login
  */
-const userLogin = async (req: IGetUserAuthRequest, res: Response) => {
-	const { email, password } = req.body;
+const userLogin = async (email: string, password: string) => {
 	try {
 		const user = await User.findOne({ email });
 		if (!user) {
 			logger.info('User not found');
```
Comment (mali-tintash) [inline src/lib/auth.ts:20]: discourage using hard coded strings.

File: src/controllers/post.ts (hunk 13 above)
Comment (Affaq-Ahmed) [reply src/controllers/post.ts:71]: Ok. Will do that in the future.

**Pull request merged**